### PR TITLE
Query Frontend: add new field for dense native histogram format

### PR DESCRIPTION
**Pull request by yeya24**
**What this PR does**:

https://github.com/cortexproject/cortex/pull/6151 added support for Ruler to evaluate rules via sending gRPC requests to Query Frontend.

An issue we found out is that the current JSON format query response from Query Frontend doesn't allow Ruler to access the whole information of native histograms. As what Prometheus returns in its Query response is a more compact format of native histogram, not its full format.

With another in progress PR https://github.com/cortexproject/cortex/pull/5527, we can make it possible to return protobuf query response in Query Frontend through Querier. This could make it possible for Ruler to get the full representation of native histograms.

This PR adds a new `rawHistograms` field to Query Frontend protobuf. This field should never be set when using JSON codec and should be only used when using Protobuf codec asked by Ruler.

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by yeya24** [comment]
@SungJin1212 Would you like to take a look? This PR tries to return dense native histogram format if it is protobuf codec with cortex internal set to true.

Then you can continue the work in https://github.com/cortexproject/cortex/pull/6345 to implement everything end to end.

File: pkg/querier/codec/protobuf_codec.go [hunk 1]
```diff
@@ -13,10 +13,20 @@ import (
...
+	CortexInternal bool
+}
 
 func (p ProtobufCodec) ContentType() v1.MIMEType {
-	return v1.MIMEType{Type: "application", SubType: "x-protobuf"}
+	if !p.CortexInternal {
+		return v1.MIMEType{Type: "application", SubType: "x-protobuf"}
+	}
+	// TODO: switch to use constants.
+	return v1.MIMEType{Type: "application", SubType: "application/x-cortex-query+proto"}
```
Comment (SungJin1212) [inline pkg/querier/codec/protobuf_codec.go:29]: subType is a `x-cortex-query+proto`
```suggestion
	eturn v1.MIMEType{Type: "application", SubType: "x-cortex-query+proto"}
```

**Review by SungJin1212 — APPROVED** [review]

**Comment by SungJin1212** [comment]
LGTM! Thanks.

File: pkg/querier/tripperware/query.proto [hunk 2]
```diff
@@ -47,7 +47,10 @@ message Extent  {
 message SampleStream {
   repeated cortexpb.LabelPair labels = 1 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "metric", (gogoproto.customtype) = "github.com/cortexproject/cortex/pkg/cortexpb.LabelAdapter"];
   repeated cortexpb.Sample samples = 2 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "values"];
-  repeated SampleHistogramPair histograms = 3 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "histograms"];
+  repeated SampleHistogramPair histograms = 3 [(gogoproto.nullable) = true, (gogoproto.jsontag) = "histograms"];
+  // Full representation of native histograms and it should be always float histogram type.
+  // This field should be never deserialized into JSON and should only exist in protobuf format.
+  repeated cortexpb.Histogram rawHistograms = 4 [(gogoproto.nullable) = true, (gogoproto.jsontag) = "-"];
```
Comment (SungJin1212) [inline pkg/querier/tripperware/query.proto:53]: I just wonder why the `rawHistograms` are needed in the `matrix` result.

Comment (SungJin1212) [reply pkg/querier/tripperware/query.proto:53]: ok, thanks.

Comment (SungJin1212) [reply pkg/querier/tripperware/query.proto]: I see. Could you merge it? Then, I start with the implementation for retrieving native histograms.

**Pull request merged**